### PR TITLE
Fix constructing GPU Tensor from DLPack capsule

### PR DESCRIPTION
**Pull request by JanuszL**
- when the DLPack capsule is provided and GPU Tensor is
  created with provided deivce_id DALI attempts to treat it as
  an object implementing __cuda_arra_interface__. This PR unifies
  the GPU Tensor constructor in the python backend and dynamically
  determine the type of provided object used by the constructor
- fixes regression from https://github.com/NVIDIA/DALI/pull/3710

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

## Category:

**Bug fix** (*non-breaking change which fixes an issue*)

## Description:
- when the DLPack capsule is provided and GPU Tensor is
  created with provided deivce_id DALI attempts to treat it as
  an object implementing __cuda_arra_interface__. This PR unifies
  the GPU Tensor constructor in the python backend and dynamically
  determine the type of provided object used by the constructor
- fixes regression from https://github.com/NVIDIA/DALI/pull/3710

### Affected modules and functionalities:
- pybind python binding
- external source

### Tests
- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [x] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

File: dali/python/nvidia/dali/external_source.py [hunk 1]
```diff
@@ -74,13 +74,13 @@ def to_numpy(x):
...
-            info = _b.CheckDLPackCapsule(datum)
-            if not info[0] and not checked:
+            (is_dlpack, is_gpu_data) = _b.CheckDLPackCapsule(datum)
+            if not is_dlpack and not checked:
                 _check_data_batch(data, batch_size, layout)
                 checked = True
             if isinstance(datum, (_tensors.TensorCPU, _tensors.TensorGPU)):
                 inp = type(datum)(datum, layout=layout) if layout is not None else datum
-            elif hasattr(datum, "__cuda_array_interface__") or (info[0] and info[1]):
+            elif hasattr(datum, "__cuda_array_interface__") or (is_dlpack and is_gpu_data):
```
Comment (mzient) [inline dali/python/nvidia/dali/external_source.py:83]: I suggest separating the code paths:
```suggestion
            elif is_dlpack:
              if is_gpu_data:
                inp = _tensors.TensorGPU(datum, layout or "")
              else:
                inp = _tensors.TensorCPU(datum, layout or "")
            elif hasattr(datum, "__cuda_array_interface__"):
```

File: dali/python/backend_impl.cc [hunk 2]
```diff
@@ -458,31 +467,20 @@ void ExposeTensor(py::module &m) {
...
-      "object"_a,
-      "layout"_a = "",
-      R"code(
-      Wrap a DLPack Tensor residing in the GPU memory.
-
-      object : DLPack object
-            Python DLPack object
-      layout : str
-            Layout of the data
-      )code")
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/python/backend_impl.cc:475]: We can check if it is passed and raise. It is pretty confusing that providing `device_id` changes how the input is treated.

File: dali/python/nvidia/dali/external_source.py (hunk 1 above)
Comment (JanuszL) [reply dali/python/nvidia/dali/external_source.py:83]: I agree with this one.

File: dali/python/backend_impl.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (mzient) [reply dali/python/backend_impl.cc:475]: Well, isn't it how function overloads are supposed to work? ;)

**Comment by JanuszL** [comment]
!build

File: dali/python/backend_impl.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/python/backend_impl.cc]: Done

**Review by mzient — APPROVED** [review]

**Review by jantonguirao — APPROVED** [review]

**Pull request merged**